Ensure immutable isbits Union fields get set correctly

quinnj opened the pull request:
@vtjnash, this isn't quite right yet. When constructed, the selector byte and any non-union fields aren't getting set correctly.

I'm kind of just taking a stab here, and I'm guessing that my use of `fval_info` and `dest` are not quite right, hence the non-working solution here. For example, when I run
```julia
struct BitsUnion
    x::Union{Int64, Float64}
end
BitsUnion(5.0)
```
the `fval_info` is
```
(lldb) call jl_(fval_info.constant)
#<null>
(lldb) call jl_(fval_info.typ)
Float64
(lldb) p fval_info.isimmutable
(const bool) $0 = false
```
I'm not sure why the `fval_info.constant` isn't set, but that causes problems when we try to `compute_tindex_unboxed` later against `BitsUnion`. 

I won't be able to do much else on this for the rest of the day, but any pointers/tips/direction you or @yuyichao could provide would be much appreciated.

fix #23351

quinnj committed:
Work to ensure immutable isbits Union fields get set correctly. #23351

quinnj commented:
Ok, a little more sleuthing:
- Turns out when you call `julia_type_to_llvm` on `Union{Int64, Float64}`, the resulting `Type *` has `isPointerTy() == true`
- I believe that's why the value setting isn't happening correctly, since `emit_unbox` thinks it's dealing w/ a pointer type when it should be unboxing inline as a `Float64` or `Int64`

I'm not quite sure what the right solution here is though. Modify `emit_unbox`? Lie to LLVM and call `emit_unbox` w/ the largest Union element? 

I feel like there still might be another issue going on, because in the llvm code for `BitsUnion(5.0)`, I still don't see a selector byte store for some reason. Would that be getting optimized out for some reason? Or "clobbered" somehow due to the Union-is-a-pointertype issue?

vtjnash commented on src/builtins.c:
```diff
@@ -259,6 +259,10 @@ JL_CALLABLE(jl_f_sizeof)
     if (jl_is_unionall(x) || jl_is_uniontype(x)) {
         x = jl_unwrap_unionall(x);
+        size_t elsize = 0, al = 0;
+        int isinline = jl_islayout_inline(x, &elsize, &al);
```
This will have the curious property that `T == S` will not imply `sizeof(T) == sizeof(S)`. I'm not sure if that's going to be an issue, but it does make be nervous. In particular, consider `Union{Tuple{Int64}, Tuple{Int8}}`, which doesn't include selector bits, vs. `Tuple{Union{Int64, Int8}}`, which throws an error.

quinnj commented:
Bump @vtjnash, I could really use your help/direction here.

vtjnash committed:
correctly handle union-splitting in codegen type allocation

quinnj commented:
Alright, added some tests for @vtjnash's help w/ the fix. Let's see what CI says. Anyone else want to take a look?

quinnj committed:
Add tests for structs with isbits Union fields.

quinnj commented:
That's a lot of great green checkmarks; a shame that Travis is backed up so far.

vtjnash commented:
Looks like Travis caught another bug though:

	From worker 5:	compare_fields at /home/travis/build/JuliaLang/julia/src/builtins.c:94
(that line is assuming isbits_layout === isdatatype)

vtjnash committed:
correctly handle union-store-splitting in "is" and "object_id"

quinnj commented:
Mac failure is #23215; everything else looks all good.

vtjnash committed:
fix codegen bugs, simplify type representations

anonymize all types, remove special representation for Ptr, fix Constant
generation for types containing split-union fields

Keno commented:
@vtjnash is this good to go?

JeffBezanson commented on src/rtutils.c:
```diff
@@ -883,9 +883,12 @@ static size_t jl_static_show_x_(JL_STREAM *out, jl_value_t *v, jl_datatype_t *vt
-                                           depth);
+                    jl_datatype_t *ft = (jl_datatype_t*)jl_field_type(vt, i);
+                    if (jl_is_uniontype(ft)) {
+                        uint8_t sel = ((uint8_t*)fld_ptr)[jl_field_size(vt, i) - 1];
```
This operation should be abstracted.

quinnj commented:
I'm still seeing a segfault w/
```julia
struct A
    x::Union{Float64, Void}
end

getx(a::A) = a.x

getx(A(1.0))
```

quinnj commented:
Ok, looks like the segfault is coming from 
```
data = ctx.builder.CreateSelect(isboxed_union,
-> 5661	                                    ctx.builder.CreateLoad(retvalinfo.gcroot),
   5662	                                    data);
```
when `retvalinfo.gcroot` is accessed. Note that I'm running master merged w/ this branch locally, in case that makes a difference.

Keno committed:
Untangle legacy .gcroot field

When the new gc root placement pass was introduced, the .gcroot field
lost its original meaning and became mostly superfluous. However, the
codegen code for unions still used it to get a referenced to the boxed
part of the split union. Clean that up by introducing an explicit field
for that purpose and removing the .gcroot field.

Keno commented:
@quinnj You might try rebasing on top of https://github.com/JuliaLang/julia/pull/23571, which cleans up some of this code. I'll take a look if it still crashes.

quinnj commented:
That indeed seems to have fixed the problem @keno. I've pushed this branch as rebased on top of #23571

quinnj commented:
Hmmm, travis linux 64-bit & julia freebsd ci both stalled during tests; I imagine unrelated? Have we seen those kinds of hangs elsewhere?

nalimilan commented:
Yes, I've seen a hang today in another PR.

quinnj commented:
Ok, rebased this again on top of @Keno 's new gcroot commit. Also added a test for the codgen-getfield crash I saw above. Let's see what CI says on all of this.

quinnj commented:
CI all looks good, just the one hang on 32-bit travis linux. I'll let @Keno & @vtjnash decide when this is good to go. I'll keep doing more testing on my end.

Keno commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -1156,10 +1156,9 @@ bool LateLowerGCFrame::CleanupIR(Function &F) {
-                    CI->getType(), "", CI);
-                ASCI->takeName(CI);
-                CI->replaceAllUsesWith(ASCI);
+                auto *ptr = new PtrToIntInst(CI->getOperand(0), CI->getType(), "", CI);
```
This isn't valid by IR invariants (it'll complain on LLVM 5.0+). Why does this need to be an int?

vtjnash replied:
That's what CI->getType() is. We can worry about LLVM 5 when we start supporting it.

Keno replied:
@yuyichao is running in that configuration AFAIK.

yuyichao replied:
I think 4.0 already has non-integral pointer so this will complain there already.

I'm on 4.0 but I do test svn from time to time to make sure it's ready whenever Arch's package is updated.

Should be enough to just keep the old `addrspacecast` and add an `ptrtoint` on top?

yuyichao replied:
This comment is not addressed and it breaks LLVM 4.0 .

Keno commented on test/libgit2.jl:
```diff
@@ -76,6 +76,8 @@ end
     a = Base.cconvert(Ptr{LibGit2.StrArrayStruct}, p)
     b = Base.unsafe_convert(Ptr{LibGit2.StrArrayStruct}, a)
     @test p == convert(Vector{String}, unsafe_load(b))
+    @noinline gcuse(a) = a
```
Should use `Core.gcuse` now.

StefanKarpinski commented:
:heart:

maleadt commented:
This breaks `llvmcall(fun::Ptr)` pretty spectacularly:


```
$ gdb julia

(gdb) break ccall.cpp:1083
Breakpoint 1 (ccall.cpp:1083) pending.

(gdb) run
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-DEV.1704 (2017-09-07 16:53 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 1875957dec* (0 days old master)
|__/                   |  x86_64-unknown-linux-gnu

julia> x() = Base.llvmcall(C_NULL, Void, Tuple{Ptr{Int}, Int}, C_NULL, 1)
x (generic function with 1 method)

julia> x()

Thread 1 "julia" hit Breakpoint 1, emit_llvmcall (ctx=..., args=0x7fffeb1c6750, nargs=5) at /home/tbesard/Work/Julia-CUDA/julia-dev/src/ccall.cpp:1083
1083            assert(isPtr);

(gdb) call argtypes[0]->dump()
i64

(gdb) call argtypes[1]->dump()
i64
```

As seen on LLVM.jl, bisected to 1875957dec25c780339d34c3a0d7fda7490b7072.

vtjnash commented:
llvmcall is still experimental (unstable) anyways

maleadt commented:
> llvmcall is still experimental (unstable) anyways

Yeah sure stuff breaks on master all the time, just reporting the fact.

maleadt commented:
OK, so looking at the actual commits this seems fully intended. Is this anonymization of pointers to native-width integers going to be the default from now on? Even though `llvmcall` is unstable, this seems unnecessarily breaking..
